cloud-language-starter/src/main/java/com/sample/autoconfig/LanguageAutoConfig.java

```diff
@@ -94,10 +70,19 @@ public TransportChannelProvider defaultLanguageTransportChannelProvider() {
     return LanguageServiceSettings.defaultTransportChannelProvider();
   }
 
+  // bean for overriding retry settings on service-level
+  @Bean
+  @ConditionalOnMissingBean(name = "languageServiceRetry")
+  public Retry languageServiceRetry() {
```

Comment:
I'm actually not sure that `Retry` needs to be a bean, as its purpose is to help define the properties object. `RetrySettings`, which is built from `LanguageProperties.retry` should probably be a bean instead.

Reply:
I think this and the following two comments are addressing a similar design decision, which I'll try to explain and ask for some opinions here! I encountered two main blockers that led to returning a `Retry` and not `RetrySettings` object in the bean: 

1. Retry settings seem to be set on a method-level and not the entire client (I was looking at [this usage](https://github.com/googleapis/java-language/blob/10af2e424caff3aa47a7470ce1520281aa881cb4/google-cloud-language/src/main/java/com/google/cloud/language/v1/stub/LanguageServiceStubSettings.java#L379) - but @blakeli0 please let me know if there is anything missed here), so it’s difficult to get existing retry defaults inside the service-level bean (in case that method-level defaults are different). 

2. If we create a `RetrySettings` object from properties to return in this bean, and update method-level settings in the `languageServiceSettings` bean, the issue is we cannot distinguish which values have been overridden through spring properties by the user. For example:
    * Looking at [RetrySettings.Builder logic](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L145), default value for gax `RetrySettings.getTotalTimeout()` is zero, suppose client library defaults this to a non-zero value, and user overrides this back to zero - in the update logic there isn't a good way to tell whether this has been configured by user or not configured at all.

Reply:
Yes, this point about service vs. method-level settings makes a lot of sense - thanks for the explanation!

My original concern with both 1) and 2) was mostly around preserving default values and the granularity of the retry settings (e.g. if user provides an override value for `totalTimeout` but not `initialRetryDelay`). But I think what I missed here is that in the service-level bean, we can try to generate code that builds a `RetrySettings` object with the defaults specified in the service config, instead of grabbing these on a method-level from an already initialized client. 

Updated with switch to `RetrySettings` bean - does this approach align with what you had in mind?

Reply:
Sorry I'm a little late to this conversation.
1. I can confirm that there is no way to set RetrySetting at service level. I thought so because we can configure a service level default in the yaml(see this [doc](https://g3doc.corp.google.com/google/g3doc/oneplatform/client-config.md?cl=head#client-side-retry-configuration) for details), so it would be weird that we don't have a way to set it in the client settings. In addition, we can also set an LRO RetrySettings at service level. This could be something we want to expose to customers in the future as well.
2.  I agree with both of you that ideally we want to have service level and method level RetrySettings. The only thing to take in mind is that we may need to design carefully when we actually want to support method level RetrySettings, as the RetrySettings are currently configured per a group of methods like [here](https://github.com/googleapis/googleapis/blob/master/google/cloud/asset/v1/cloudasset_grpc_service_config.json), you can also configure a service level RetrySetting, combine it with customer specified service level and method level RetrySettings, the code could become messy and error prone.

Reply:
> The only thing to take in mind is that we may need to design carefully when we actually want to support method level RetrySettings, as the RetrySettings are currently configured per a group of methods like [here](https://github.com/googleapis/googleapis/blob/master/google/cloud/asset/v1/cloudasset_grpc_service_config.json)

@blakeli0 Thank you for the cloudasset example! I did not realize these configs are per-[groups of methods] instead of per-service, so there probably isn’t always a set of service-level defaults configured by the client library. I think this breaks my attempt on the service-level `RetrySettings` bean, if we still intend to fall back to client-library-configured defaults. 

> Once you have a RetrySettings bean, why doesn't matter how it was constructed, whether it's properties or user-defined? You simply assign it to the method settings.

@meltsufin The problem I had with a `RetrySettings` bean on service-level is the preservation of defaults for fields that the user does not intend to override. I pushed a new commit to compare this against a method-level bean, but here is a complication scenario:
- When a `RetrySettings` object is built, the builder [assigns default values](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L145) (e.g. timeout 0s, multiplier 1.0) to all fields so that they are not null.
- On the client library side, there are [default retry settings configured](https://github.com/googleapis/googleapis/blob/f3d6f41ed50ae9271fbf9ce2355d1f1e8afb4d94/google/cloud/language/v1/language_grpc_service_config.json#L30) at method level - (e.g. timeout 0.1s, multiplier 1.3).
- Suppose that in the default service-level bean, we build a `RetrySettings` object from properties (similar to [in pubsub](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/1f9190d4bdd80df17fd9e7c6ac95ea9f07b538f0/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/GcpPubSubAutoConfiguration.java#L308)), but the properties only configure timeout (= 0.2s) and not multiplier. In this object multiplier would take on a builder default of 1.0. 
- When we use this object later to set method-level retry settings,  we can either 1) replace the RetrySettings object entirely, overriding all method-level defaults from the client library, or 2) compare against the method-level retry settings and only update fields that have been custom-configured. 
- But for 2), we cannot identify which subset of fields have been updated in a RetrySettings object (e.g. if user sets multiplier back to 0.0 through properties, this coincides with the builder default value)  -  this is what I was trying to work around earlier with using a `Retry` bean instead.

Reply:
1. I am a bit hesitant to set default values explicitly in below code. I moved away from this in original implementation since this would mean that our starters are more client library version specific than necessary -- a mismatch in client library version user uses may result in unexpected behavior. One other concern is that client library may not have defaults set (vision [config](https://github.com/googleapis/googleapis/blob/master/google/cloud/vision/v1/vision_grpc_service_config.json)) on service level but have settings on method level, in that case do we retrieve the gax defaults?
https://github.com/zhumin8/language_poc1/blob/4b05ac3ab7f03130b37eda9afa3aaa1472becb19/cloud-language-starter/src/main/java/com/sample/autoconfig/LanguageAutoConfig.java#L79-L85

2. I like the idea to have both service level and method level RetrySettings too. However, since client library itself only supports setting on method level, I think it is reasonable to start with only method level, and keep service level as a future enhancement. This could also be easier since we can retrieve client library defaults like ` LanguageServiceSettings.newBuilder().annotateTextSettings().getRetrySettings()`.

What do you all think?

Reply:
Have you considered the `RetrySettings.merge()` method? This allows you to have a `RetrySettings` bean at either or both service-level and method-level, and then merge it into each of the method-level retry settings, letting GAPIC handle the defaults. Can we just start with a service-level retry properties that we turn into a service level `RetrySettings` bean, and then apply it to each of the method-level retry settings using ` LanguageServiceSettings.newBuilder().[method]Settings().retrySettings().merge()`?
We can of course just do method-level settings as well. Whatever makes more sense, but I don't really see the issue with handling defaults. If you had both service and method-level settings, you would apply them by calling `LanguageServiceSettings.newBuilder().[method]Settings().retrySettings().merge(serviceLevelSettings).merge(methodLevelSettings)`.

Reply:
Right - the issue here is that `RetrySettings.merge()` doesn't work as expected in our use case: 
[these non-null check conditions](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L349) will always be entered, because the `RetrySettings` object is [built with defaults](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L145) such as `Duration.ZERO`. So we will always be overwriting all GAPIC defaults with GAX defaults with `merge()`. 

Maybe the root solution here is a change in GAX `RetrySettings` code. But for now, I’m aligned with @zhumin8 ’s assessment that it’s easier to work with defaults for method-level beans - I'll give the implementation for that a shot, and we can decide whether to keep or exclude it from the preview.

Reply:
I see, but what if you set everything in the builder to `null` before assigning values from the properties?

Reply:
Currently the fields with primitive type (`int` and `double`) in `RetrySettings` cannot be null (whereas the `Retry` object uses `Integer` and `Double` types to allow for this)

Reply:
Right, but the merge method logic has rules for which values of the primitive types are considered "empty".

Reply:
It looks like the rules there specify more "invalid" than "empty" values. For example, `build()` [throws an exception](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L328) if multipliers (double) are set to < 1.0, and the merge logic checks for multipliers >= 1. 

We can set everything in the builder to these "invalid" values, but cannot build it into a `RetrySettings` object to return.

Reply:
Can we use `RetrySettings.Builder` for the beans instead of `RetrySettings`?

Reply:
Hmm, that does allow us to work with most of the `merge()` logic, except for the [`maxAttempts` field](https://github.com/googleapis/gax-java/blob/5e76b4b4487d90623dd8daa0b0249505fa97c99d/gax/src/main/java/com/google/api/gax/retrying/RetrySettings.java#L361) which is always overwritten. I'll have to give it some more thought.

Revisiting the start of this thread: would the usage of a `Retry` bean make sense given this limitation, or is there a reason we'd prefer something like `RetrySettings.Builder` over it?

Reply:
I think we can also just not expose retry settings as beans, and just stick with properties for now. If the user wants to provide their own bean, they would need to do it by providing the entire ClientSettings bean. In the future, this can be expanded to be more granular to allow providing retry settings beans. BTW, we can also expose the ClientSettings.Builder as a bean, and users can then apply their own bean post processor to adjust the settings.

Reply:
Configuring service-level retry settings through properties sounds like a good MVP! I’ll go ahead with this, and will put together a summary for future enhancement options (exposing beans, method-level configurations) and design alternatives discussed.